File: System.Drawing/System.Drawing/Graphics.cs
// Copyright (c) 2009 DotNetAnywhere
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;
using System.Drawing.Text;

namespace System.Drawing {
	public sealed class Graphics : MarshalByRefObject, IDisposable {

		public static Graphics FromHdc(IntPtr hdc) {
			if (hdc != IntPtr.Zero) {
				throw new NotImplementedException();
			}
			int xSize, ySize;
			PixelFormat pixelFormat;
			IntPtr native = LibIGraph.GetScreen(out xSize, out ySize, out pixelFormat);
			if (native == IntPtr.Zero) {
				throw new Exception("CreateScreen() failed");
			}
			return new Graphics(xSize, ySize, pixelFormat, native);
		}

		public static Graphics FromImage(Image image) {
			if (image == null) {
				throw new ArgumentNullException("image");
			}
			IntPtr native = LibIGraph.GetGraphicsFromImage(image.native);
			if (native == IntPtr.Zero) {
				throw new ArgumentException("Failed to get Graphics from given Image", "image");
			}
			return new Graphics(image.Width, image.Height, image.PixelFormat, native);
		}

		private IntPtr native = IntPtr.Zero;
		private int xSizePixels, ySizePixels;
		private PixelFormat pixelFormat;
		private TextRenderingHint textRenderingHint;
		private Region clip;

		private Graphics(int xSize, int ySize, PixelFormat pixelFormat, IntPtr native) {
			this.xSizePixels = xSize;
			this.ySizePixels = ySize;
			this.pixelFormat = pixelFormat;
			this.native = native;
			this.textRenderingHint = TextRenderingHint.SystemDefault;
			this.Clip = new Region();
		}

		~Graphics() {
			this.Dispose();
		}

		public void Dispose() {
			if (this.native != IntPtr.Zero) {
				LibIGraph.DisposeGraphics(this.native);
				this.native = IntPtr.Zero;
				GC.SuppressFinalize(this);
			}
		}

		public Region Clip {
			get {
				return this.clip;
			}
			set {
				if (value == null) {
					throw new ArgumentNullException();
				}
				this.clip = value;
				LibIGraph.Graphics_SetClip(this.native, value.native);
			}
		}

		public void ResetClip() {
			this.Clip = new Region();
		}

		public void Clear(Color color) {
			LibIGraph.Clear(this.native, color.ToArgb());
		}

		public TextRenderingHint TextRenderingHint {
			get {
				return this.textRenderingHint;
			}
			set {
				LibIGraph.TextRenderingHint_Set(this.native, value);
				this.textRenderingHint = value;
			}
		}

		public void DrawLine(Pen pen, int x1, int y1, int x2, int y2) {
			LibIGraph.DrawLine_Ints(this.native, pen.native, x1, y1, x2, y2);
		}

		public void DrawLine(Pen pen, Point pt1, Point pt2) {
			LibIGraph.DrawLine_Ints(this.native, pen.native, pt1.X, pt1.Y, pt2.X, pt2.Y);
		}

		public void DrawLines(Pen pen, Point[] points) {
			int num = points.Length - 1;
			for (int i = 0; i < num; i++) {
				this.DrawLine(pen, points[i], points[i + 1]);
			}
		}

		public void DrawPolygon(Pen pen, Point[] points) {
			this.DrawLines(pen, points);
			this.DrawLine(pen, points[points.Length - 1], points[0]);
		}

		public void DrawRectangle(Pen pen, int x, int y, int width, int height) {
			DrawLines(pen, new Point[] {
				new Point(x, y),
				new Point(x + width, y),
				new Point(x + width, y + height),
				new Point(x, y + height),
				new Point(x, y)
			});
		}

		public void DrawRectangle(Pen pen, Rectangle rect) {
			this.DrawRectangle(pen, rect.X, rect.Y, rect.Width, rect.Height);
		}

		public void DrawRectangle(Pen pen, float x, float y, float width, float height) {
			this.DrawRectangle(pen, (int)x, (int)y, (int)width, (int)height);
		}

		public void FillRectangle(Brush brush, int x, int y, int width, int height) {
			LibIGraph.FillRectangle_Ints(this.native, brush.native, x, y, width, height);
		}

		public void FillRectangle(Brush brush, Rectangle rect) {
			this.FillRectangle(brush, rect.X, rect.Y, rect.Width, rect.Height);
		}

		public void FillRectangle(Brush brush, float x, float y, float width, float height) {
			this.FillRectangle(brush, (int)x, (int)y, (int)width, (int)height);
		}

		public void FillRectangle(Brush brush, RectangleF rect) {
			this.FillRectangle(brush, rect.X, rect.Y, rect.Width, rect.Height);
		}

		public void DrawEllipse(Pen pen, int x, int y, int width, int height) {
			LibIGraph.DrawEllipse_Ints(this.native, pen.native, x, y, width, height);
		}

		public void DrawEllipse(Pen pen, Rectangle rect) {
			this.DrawRectangle(pen, rect.X, rect.Y, rect.Width, rect.Height);
		}

		public void FillEllipse(Brush brush, int x, int y, int width, int height) {
			LibIGraph.FillEllipse_Ints(this.native, brush.native, x, y, width, height);
		}

		public void FillEllipse(Brush brush, Rectangle rect) {
			this.FillEllipse(brush, rect.X, rect.Y, rect.Width, rect.Height);
		}

		public void DrawString(string s, Font font, Brush brush, float x, float y, StringFormat format) {
			LibIGraph.DrawString(this.native, s, font.native, brush.native, (int)x, (int)y, int.MaxValue, int.MaxValue, (format == null) ? IntPtr.Zero : format.native);
		}

		public void DrawString(string s, Font font, Brush brush, float x, float y) {
			this.DrawString(s, font, brush, x, y, StringFormat.GenericDefault);
		}

		public void DrawString(string s, Font font, Brush brush, PointF point) {
			this.DrawString(s, font, brush, point.X, point.Y, StringFormat.GenericDefault);
		}

		public void DrawString(string s, Font font, Brush brush, RectangleF rect) {
			LibIGraph.DrawString(this.native, s, font.native, brush.native,
				(int)rect.Left, (int)rect.Top, (int)rect.Right, (int)rect.Bottom,
				StringFormat.GenericDefault.native);
		}

		public void DrawString(string s, Font font, Brush brush, RectangleF rect, StringFormat format) {
			LibIGraph.DrawString(this.native, s, font.native, brush.native,
				(int)rect.Left, (int)rect.Top, (int)rect.Right, (int)rect.Bottom, format.native);
		}

		public SizeF MeasureString(string s, Font font) {
			return this.MeasureString(s, font, int.MaxValue, StringFormat.GenericDefault);
		}

		public SizeF MeasureString(string s, Font font, int width) {
			return this.MeasureString(s, font, width, StringFormat.GenericDefault);
		}

		public SizeF MeasureString(string s, Font font, int width, StringFormat format) {
			int szWidth, szHeight;
			LibIGraph.MeasureString(this.native, s, font.native, width, format.native, out szWidth, out szHeight);
			return new SizeF(szWidth, szHeight);
		}

		public void DrawImage(Image image, int x, int y) {
			LibIGraph.DrawImageUnscaled(this.native, image.native, x, y);
		}

		public void DrawImageUnscaled(Image image, int x, int y) {
			LibIGraph.DrawImageUnscaled(this.native, image.native, x, y);
		}

		public void CopyFromScreen(int srcX, int srcY, int destX, int destY, Size size) {
			LibIGraph.Graphics_CopyFromScreen(this.native, srcX, srcY, destX, destY, size.Width, size.Height);
		}

		public void CopyFromScreen(Point upperLeftSource, Point upperLeftDestination, Size size) {
			this.CopyFromScreen(upperLeftSource.X, upperLeftSource.Y, upperLeftDestination.X, upperLeftDestination.Y, size);
		}
	}
}

File: libIGraph/Graphics.c
// Copyright (c) 2012 DotNetAnywhere
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "libIGraph.h"

#include "Graphics.h"
#include "GetScreen.h"
#include "Pixels.h"
#include "Image.h"

tGraphics* InitGraphics() {
	tGraphics *pGraphics = TMALLOC(tGraphics);

	memset(pGraphics, 0, sizeof(tGraphics));

	return pGraphics;
}

void DisposeGraphics_(tGraphics *pGraphics) {
	if (pGraphics->screenPtr != NULL) {
		// This is the screen window
		ScreenDispose();
	} else {
		// This is an bitmap of some kind
	}
}

void TextRenderingHint_Set_(tGraphics *pGraphics, U32 textRenderingHint) {
	pGraphics->textRenderingHint = textRenderingHint;
}

static void Clear_WithClipping(tGraphics *pGraphics, U32 col) {
	tRegion *pClip = pGraphics->pClip;
	U32 x0 = (pClip->x1 > 0) ? pClip->x1 : 0;
	U32 x1 = (pClip->x2 < pGraphics->xSize) ? pClip->x2 : pGraphics->xSize;
	U32 y0 = (pClip->y1 > 0) ? pClip->y1 : 0;
	U32 y1 = (pClip->y2 < pGraphics->ySize) ? pClip->y2 : pGraphics->ySize;
	U32 x, y;
	for (y = y0; y<y1; y++) {
		for (x = x0; x<x1; x++) {
			mSetPixelNoBlend(pGraphics, x, y, col);
		}
	}
}

static void Clear_FMT_4BPP_GRAY(tGraphics *pGraphics, U32 col) {
	if (!pGraphics->pClip->isInfinite) {
		Clear_WithClipping(pGraphics, col);
	} else {
		U8 col4 = COL2GRAY(col) >> 4;
#ifdef INVERT_4BPP_GRAY
		col4 = 15 - col4;
#endif
		memset(pGraphics->pScan0, col4 | (col4 << 4), pGraphics->memSize);
	}
}

static void Clear_FMT_32BPP_ARGB(tGraphics *pGraphics, U32 col) {
	if (!pGraphics->pClip->isInfinite) {
		Clear_WithClipping(pGraphics, col);
	} else {
		void *pEnd = ((U8*)pGraphics->pScan0) + pGraphics->stride * pGraphics->ySize;
		U32 *pMem = (U32*)pGraphics->pScan0;
		while ((void*)pMem < pEnd) {
			*pMem++ = col;
		}
	}
}

tClear mClear_[FMT_NUM] = {
	Clear_FMT_4BPP_GRAY,
	Clear_FMT_32BPP_ARGB
};

void DrawImageUnscaled_(tGraphics *pGraphics, tImage *pImage, I32 x, I32 y) {
	I32 xPos, yPos;
	U32 col;

	// TODO: Optimize!!!
	for (yPos = 0; yPos < pImage->height; yPos++) {
		for (xPos = 0; xPos < pImage->width; xPos++) {
			col = mImage_GetPixel(pImage, xPos, yPos);
			mSetPixel(pGraphics, x + xPos, y + yPos, col);
		}
	}
}

void Graphics_SetClip_(tGraphics *pGraphics, tRegion *pRegion) {
	pGraphics->pClip = pRegion;
}

void Graphics_CopyFromScreen_(tGraphics *pGraphics, I32 srcX, I32 srcY, I32 destX, I32 destY, I32 sizeX, I32 sizeY) {
	I32 x, y, dx, dy;
	I32 xEnd = srcX + sizeX;
	I32 yEnd = srcY + sizeY;

	// TODO: Optimize!!!
	for (y=srcY, dy=destY; y<yEnd; y++, dy++) {
		for (x=srcX, dx=destX; x<xEnd; x++, dx++) {
			U32 col = GetScreenPixel(x, y);
			mSetPixel(pGraphics, dx, dy, col);
		}
	}
}
